### PR TITLE
BadPacketsH - Enforce the attack packet order (1.7 - 1.18.2)

### PR DESCRIPTION
**Pull request by LewUwU**

### CODE REVIEW

**Comment by LewUwU** [comment]
This is a impossible to flag check unless you modify the client

Packet Order is guaranteed due to TCP

**Comment by amnoah** [comment]
First off, switch statements exist.
Second off, this isn't that great. While yes, it will catch noswing cheats in 1.8 clients only, you could make it compatible for all versions.

If the player has sent an Animation packet, set swung to true. Once they attack, flag is swung is false and then reset swung to false.

This is much simpler, doesn't require you to limit versions, can still catch the cheats, and will perform identically in real scenarios (after the first hit it'll guaranteed catch the cheat. During combat you don't typically start digging so there are no other ways to send Animation).

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package ac.grim.grimac.checks.impl.badpackets;
+
+import ac.grim.grimac.checks.CheckData;
+import ac.grim.grimac.checks.type.PacketCheck;
+import ac.grim.grimac.player.GrimPlayer;
+import com.github.retrooper.packetevents.event.PacketReceiveEvent;
+import com.github.retrooper.packetevents.protocol.packettype.PacketType;
+import com.github.retrooper.packetevents.wrapper.play.client.WrappedPacketInUseEntity;
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java:8]: WrappedPacketInUseEntity is the class name in PacketEvents 1.8
```suggestion
import com.github.retrooper.packetevents.wrapper.play.client.WrapperPlayClientInteractEntity;
```

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public class BadPacketsH extends PacketCheck {
+    private boolean swung;
+
+    public BadPacketsH(final GrimPlayer player) {
+        super(player);
+    }
+
+    @Override
+    public void onPacketReceive(PacketReceiveEvent event) {
+        if (event.getPacketType() == PacketType.Play.Client.USE_ENTITY) {
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java:21]: Should be INTERACT_ENTITY. USE_ENTITY is the property on PacketEvents 1.8
```suggestion
        if (event.getPacketType() == PacketType.Play.Client.INTERACT_ENTITY) {
```

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            WrappedPacketInUseEntity packet = new WrappedPacketInUseEntity(event);
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java:22]:
```suggestion
            WrapperPlayClientInteractEntity packet = new WrapperPlayClientInteractEntity(event);
```

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+            if (packet.getAction() != WrappedPacketInUseEntity.EntityUseAction.ATTACK) return;
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java:24]:
```suggestion
            if (packet.getAction() != WrapperPlayClientInteractEntity.InteractAction.ATTACK) return;
```

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+			
+	    if (!swung) flagAndAlert();
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsH.java:26]: Incorrect indentation

**Review by Jaren8r — CHANGES_REQUESTED** [review]
This looks like you tried to copy code from another anti cheat running an older version of PacketEvents without verifying if it compiles correctly.

**Review by Jaren8r — CHANGES_REQUESTED** [review]
While the code looks good, this check false positives on newer minecraft versions because the attack is sent before the animation.

**Comment by LewUwU** [comment]
Tested, this will work only for bad coded cheats like poorly made Criticals in packet mode and etc

**Review by Jaren8r — APPROVED** [review]
Looks good for detecting NoSwing

**Comment by MWHunter** [comment]
@amnoah is this check valid?

**Comment by amnoah** [comment]
> 

Yes it should work, although it isn't necessarily the best way possible.

**Pull request merged**